crates/services/txpool_v2/src/tests/pool.rs

```diff
@@ -0,0 +1,1256 @@
+    universe.build_pool();
+    let contract_id = Contract::EMPTY_CONTRACT_ID;
+
+    let gas_coin = universe.setup_coin().1;
+    let tx = TransactionBuilder::create(
+        Default::default(),
+        Default::default(),
+        Default::default(),
+    )
+    .add_input(gas_coin)
+    .add_output(create_contract_output(contract_id))
+    .finalize_as_transaction();
+    let utxo_id = UtxoId::new(tx.id(&Default::default()), 0);
+
+    // Given
+    // create a second transaction with utxo id referring to
+    // the wrong type of utxo (contract instead of coin)
+    let random_predicate =
+        universe.random_predicate(AssetId::BASE, TEST_COIN_AMOUNT, Some(utxo_id));
+    let tx_faulty =
+        universe.build_script_transaction(Some(vec![random_predicate]), None, 0);
+
+    // When
+    let result1 = universe.verify_and_insert(tx).await;
+    let result2 = universe.verify_and_insert(tx_faulty).await;
+
+    // Then
+    assert!(result1.is_ok());
+    let err = result2.unwrap_err();
+
+    assert!(matches!(err, Error::UtxoNotFound(id) if id == utxo_id));
+}
+
+#[tokio::test]
+async fn insert__already_known_tx() {
+    let mut universe = TestPoolUniverse::default().config(Config {
+        utxo_validation: false,
+        ..Default::default()
+    });
+    universe.build_pool();
+
+    // Given
+    let tx = universe.build_script_transaction(None, None, 0);
+
+    // When
+    let result1 = universe.verify_and_insert(tx.clone()).await;
+    let result2 = universe.verify_and_insert(tx).await;
+
+    // Then
+    assert!(result1.is_ok());
+    let err = result2.unwrap_err();
+    assert!(matches!(err, Error::DuplicateTxId(_)));
+}
+
+#[tokio::test]
+async fn insert__unknown_utxo() {
+    let mut universe = TestPoolUniverse::default();
+    universe.build_pool();
+
+    // Given
+    let input = universe.random_predicate(AssetId::BASE, TEST_COIN_AMOUNT, None);
+    let utxo_id = input.utxo_id().cloned().unwrap();
+    let tx = universe.build_script_transaction(Some(vec![input]), None, 0);
+
+    // When
+    let result = universe.verify_and_insert(tx).await;
+
+    // Then
+    let err = result.unwrap_err();
+    assert!(matches!(err, Error::UtxoNotFound(id) if id == utxo_id));
+}
+
+#[tokio::test]
+async fn insert_higher_priced_tx_removes_lower_priced_tx() {
+    let mut universe = TestPoolUniverse::default();
+    universe.build_pool();
+
+    // Given
+    let common_coin = universe.setup_coin().1;
+    let tx1 =
+        universe.build_script_transaction(Some(vec![common_coin.clone()]), None, 10);
+    let tx_id = tx1.id(&ChainId::default());
+    let tx2 = universe.build_script_transaction(Some(vec![common_coin]), None, 20);
+
+    // When
+    universe.verify_and_insert(tx1).await.unwrap();
+    let result = universe.verify_and_insert(tx2).await.unwrap();
+
+    // Then
+    assert_eq!(result[0].id(), tx_id);
+}
+
+#[tokio::test]
+async fn insert__colliding_dependent_underpriced() {
```

Comment:
```suggestion
async fn insert__colliding_dependent_and_underpriced_returns_error() {
```

Does this name make more sense? I'm trying to make sure I understand.

Reply:
Yes, done.